src/Sarif/PeekableStream.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+
+namespace Microsoft.CodeAnalysis.Sarif.Readers
+{
+    /// <summary>
+    /// This stream provides a very limited Seekability for Streams which would benefit from an ersatz Seek(Origin)
+    /// capability, but which only need it at the beginning of a Stream.  Ie: to categorize binary versus textual 
+    /// data.
+    /// </summary>
+    internal class PeekableStream : Stream
+    {
+        private readonly Stream underlyingStream;
+        private byte[] rewindBuffer;
+        private int cursor;
+        public PeekableStream(Stream underlyingStream, int initialPeekWindow)
+        {
+            this.underlyingStream = underlyingStream;
+            this.rewindBuffer = new byte[initialPeekWindow];
+            this.cursor = 0;
+        }
+
+        public override bool CanRead => true;
+
+        // Rewind() is much weaker than Seek.  We should not advertise a capability we do not have.
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => false;
+
+        public override long Length => underlyingStream.Length;
+
+        public override long Position 
+        {
+            get => Math.Min(underlyingStream.Position, cursor);
+            set => throw new System.NotImplementedException();
+        }
+
+        public override void Flush()
+        {
+            throw new System.NotImplementedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (this.rewindBuffer == null)
+            {
+                // Common case: we're past the "peekability window" and are now just a thin wrapper around the underlying Stream.
+                return underlyingStream.Read(buffer, offset, count);
+            }
+            else if (this.cursor == this.underlyingStream.Position)
+            {
+                // Buffer cursor is "caught up" to stream position, so we are not in a "rewound" state.
+                if (this.cursor == this.rewindBuffer.Length)
+                {
+                    // A previous read has brought us to the brink of exceeding the "peekability window", and we are
+                    // now reading more.  So we abandon the buffer, and serve the read as a thin wrapper.
+                    this.rewindBuffer = null;
+                    this.cursor = int.MaxValue;
+                    return underlyingStream.Read(buffer, offset, count);
+                }
+                else
```

Comment:
>else [](http://example.com/codeflow?start=16&length=4)

Again, I'd elide this `else`. when your previous if statement returns, just get rid of it. it makes it *much* clearer that there's no possibility of a sneaky return at the end of this `else` clause. Also, doing this left shifts a significant chunk of code, and less nesting == more readable, again IMO